src/ios/run.ts

```diff
@@ -10,13 +10,94 @@ import {
   IOSRunException,
 } from '../errors';
 import { getOptionValue } from '../utils/cli';
+import { wait } from '../utils/process';
 
+import type { DeviceValues } from './lib';
+import { IOSLibError } from './lib/lib-errors';
 import { getBundleId, unzipIPA } from './utils/app';
 import { getConnectedDevices, runOnDevice } from './utils/device';
+import type { SimulatorResult } from './utils/simulator';
 import { getSimulators, runOnSimulator } from './utils/simulator';
 
 const debug = Debug('native-run:ios:run');
 
+interface IOSRunConfig {
+  udid?: string;
+  devices: DeviceValues[];
+  simulators: SimulatorResult[];
+  appPath: string;
+  bundleId: string;
+  waitForApp: boolean;
+  preferSimulator: boolean;
+}
+
+async function runIpaOrAppFile({
+  udid,
+  devices,
+  simulators,
+  appPath,
+  bundleId,
+  waitForApp,
+  preferSimulator,
+}: IOSRunConfig): Promise<void> {
+  if (udid) {
+    if (devices.find(d => d.UniqueDeviceID === udid)) {
+      await runOnDevice(udid, appPath, bundleId, waitForApp);
+    } else if (simulators.find(s => s.udid === udid)) {
+      await runOnSimulator(udid, appPath, bundleId, waitForApp);
+    } else {
+      throw new IOSRunException(
+        `No device or simulator with UDID "${udid}" found`,
+        ERR_TARGET_NOT_FOUND,
+      );
+    }
+  } else if (devices.length && !preferSimulator) {
+    // no udid, use first connected device
+    await runOnDevice(devices[0].UniqueDeviceID, appPath, bundleId, waitForApp);
+  } else {
+    // use default sim
+    await runOnSimulator(
+      simulators[simulators.length - 1].udid,
+      appPath,
+      bundleId,
+      waitForApp,
+    );
+  }
+}
+
+async function runIpaOrAppFileOnInterval(config: IOSRunConfig): Promise<void> {
+  const retryInterval = 5000;
+  let error: Error | undefined;
+
+  const retry = async () => {
+    console.log('Device Locked - Checking for unlock in 5 seconds');
+    await wait(retryInterval);
+    await run();
+  };
+
+  const run = async () => {
+    try {
+      await runIpaOrAppFile(config);
+    } catch (err) {
+      if (err instanceof IOSLibError && err.code == 'DeviceLocked') {
+        await retry();
+      } else {
+        error = err;
+      }
+    }
+  };
+
+  return new Promise((resolve, reject) => {
+    return run().then(() => {
+      if (error) {
+        reject(error);
+      } else {
+        resolve();
+      }
+    });
+  });
```

Comment:
Should be able to simplify this with async/await. (But please test.)

```suggestion
  await run();
  
  if (error) {
    throw error;
  }
```

Reply:
Yup this does indeed work